nessie/src/main/java/org/apache/iceberg/nessie/NessieIcebergClient.java

```diff
@@ -181,133 +186,220 @@ public IcebergTable table(TableIdentifier tableIdentifier) {
-                      .namespace(org.projectnessie.model.Namespace.of(namespace.levels())))
-              .get();
-      return response.getNamespaces().stream()
-          .map(ns -> Namespace.of(ns.getElements().toArray(new String[0])))
-          .filter(ns -> ns.length() == namespace.length() + 1)
+      org.projectnessie.model.Namespace root =
+          org.projectnessie.model.Namespace.of(namespace.levels());
+      String filter =
+          namespace.isEmpty()
+              ? "size(entry.keyElements) == 1"
+              : String.format(
+                  "size(entry.keyElements) == %d && entry.encodedKey.startsWith('%s.')",
+                  root.getElementCount() + 1, root.name());
+      List<ContentKey> entries =
+          withReference(api.getEntries()).filter(filter).stream()
+              .filter(e -> Content.Type.NAMESPACE.equals(e.getType()))
+              .map(EntriesResponse.Entry::getName)
+              .collect(Collectors.toList());
+      if (entries.isEmpty()) {
+        return Collections.emptyList();
+      }
+      GetContentBuilder getContent = withReference(api.getContent());
+      entries.forEach(getContent::key);
+      return getContent.get().values().stream()
+          .map(v -> v.unwrap(org.projectnessie.model.Namespace.class))
+          .filter(Optional::isPresent)
+          .map(Optional::get)
+          .map(v -> Namespace.of(v.getElements().toArray(new String[0])))
+          .filter(v -> v.length() == namespace.length() + 1) // only direct children
           .collect(Collectors.toList());
-    } catch (NessieReferenceNotFoundException e) {
+    } catch (NessieNotFoundException e) {
+      if (namespace.isEmpty()) {
+        throw new NoSuchNamespaceException(
+            e,
+            "Cannot list top-level Namespaces: ref '%s' is no longer valid.",
+            getRef().getName());
+      }
       throw new RuntimeException(
           String.format(
-              "Cannot list Namespaces starting from '%s': " + "ref '%s' is no longer valid.",
+              "Cannot list child Namespaces from '%s': ref '%s' is no longer valid.",
               namespace, getRef().getName()),
           e);
     }
   }
 
   public boolean dropNamespace(Namespace namespace) throws NamespaceNotEmptyException {
+    getRef().checkMutable();
+    ContentKey key = ContentKey.of(namespace.levels());
     try {
-      getRef().checkMutable();
-      withReference(
-              getApi()
-                  .deleteNamespace()
-                  .namespace(org.projectnessie.model.Namespace.of(namespace.levels())))
-          .delete();
-      refresh();
-      return true;
-    } catch (NessieNamespaceNotFoundException e) {
-      return false;
+      Map<ContentKey, Content> contentMap =
+          api.getContent().reference(getReference()).key(key).get();
+      Content existing = contentMap.get(key);
+      if (existing != null && !(existing instanceof org.projectnessie.model.Namespace)) {
```

Comment:
```suggestion
      if (existing != null && !existing.getType().equals(Content.Type.NAMESPACE)) {
```

Reply:
I think this is slightly better to read because you don't need to wrap the whole statement in () for negation